include/alpaka/mem/buf/BufOacc.hpp

```diff
@@ -0,0 +1,652 @@
+                    {
+                        pitchBytes[i-1] = extent[i-1] * pitchBytes[i];
+                    }
+                    return pitchBytes;
+                }
+
+            public:
+                //-----------------------------------------------------------------------------
+                //! Constructor
+                template<
+                    typename TExtent>
+                ALPAKA_FN_HOST BufOaccImpl(
+                    DevOacc const & dev,
+                    TElem * const pMem,
+                    TExtent const & extent) :
+                        m_dev(dev),
+                        m_extentElements(extent::getExtentVecEnd<TDim>(extent)),
+                        m_pitchBytes(calculatePitchesFromExtents(m_extentElements)),
+                        m_pMem(pMem)
+                {
+                    ALPAKA_DEBUG_MINIMAL_LOG_SCOPE;
+
+                    static_assert(
+                        TDim::value == alpaka::Dim<TExtent>::value,
+                        "The dimensionality of TExtent and the dimensionality of the TDim template parameter have to be identical!");
+                    static_assert(
+                        std::is_same<TIdx, Idx<TExtent>>::value,
+                        "The idx type of TExtent and the TIdx template parameter have to be identical!");
+                }
+
+            public:
+                DevOacc m_dev;
+                Vec<TDim, TIdx> m_extentElements;
+                Vec<TDim, TIdx> m_pitchBytes;
+                TElem* m_pMem;
+
+                    BufOaccImpl(const BufOaccImpl&) = delete;
+                    BufOaccImpl(BufOaccImpl &&) = default;
+                BufOaccImpl& operator=(const BufOaccImpl&) = delete;
+                BufOaccImpl& operator=(BufOaccImpl&&) = default;
+                    ~BufOaccImpl()
+                {
+                    m_dev.makeCurrent();
+                    acc_free(m_pMem);
+                }
+            };
+        }
+    }
+
+    template<
+        typename TElem,
+        typename TDim,
+        typename TIdx>
+    class BufOacc
+    {
+    public:
+        //-----------------------------------------------------------------------------
+        //! Constructor
+        template<
+            typename TExtent>
+        ALPAKA_FN_HOST BufOacc(
+            DevOacc const & dev,
+            TElem * const pMem,
+            TExtent const & extent) :
+                m_spBufImpl(std::make_shared<oacc::detail::BufOaccImpl<TElem, TDim, TIdx>>(dev, pMem, extent))
+        {}
+
+            BufOacc(const BufOacc&) = default;
+            BufOacc(BufOacc &&) = default;
+        BufOacc& operator=(const BufOacc&) = default;
+        BufOacc& operator=(BufOacc&&) = default;
+
+        oacc::detail::BufOaccImpl<TElem, TDim, TIdx>& operator*() {return *m_spBufImpl;}
+        const oacc::detail::BufOaccImpl<TElem, TDim, TIdx>& operator*() const {return *m_spBufImpl;}
```

Comment:
Using `operator*` seems like a weird thing to do. In what context will this be used?

Reply:
@jkelling

Reply:
This is basically a getter to access the `Impl` instance as a reference. In a way, this is a pointer type as it is implicitly shared.

The way this is implemented here is the same as in the `BufOmp5`. It is different from `BufCpu` only in that the impl pointer is private which I find to be cleaner than having it public. These operators are only used by trait specializations and for these it felt more explicit to do the `(*buf).`-thing, then use `buf->` when I die this for `BufOmp5`.

A clean implementation of an implicitly shared type would not expose the impl pointer at all and instead wrap all the impl's functionality with member functions or have the traits as friends (the former being more clean.). I think there was a discussion on this before. This is not done in the clean way in alpaka currently, probably because of the amounts of boiler-plate involved.

I suggest to leave this as-is for now and maybe open an against both `BufOmp5` and `BufOacc` (maybe also `BufCpu`) to clean this up.

Reply:
I feel like we should revisit the syntax at some point. `operator*` does not feel like the right choice here. Especially, if a corresponding `operator->` is not provided. But then I also do not understand the shared buffer implementation well enough. Leave it as is for now.

Reply:
Opened issue https://github.com/alpaka-group/alpaka/issues/1207 .